src/get-started/create-service.md

```diff
@@ -440,94 +418,64 @@ We want to implement 2 read requests:
+    fn get_wallet(state: &ServiceApiState, query: WalletQuery)
+     -> api::Result<Wallet> {
+        let snapshot = state.snapshot();
         let schema = CurrencySchema::new(snapshot);
-
-        if let Some(wallet) = schema.wallet(&public_key) {
-            self.ok_response(&serde_json::to_value(wallet).unwrap())
-        } else {
-            self.not_found_response(
-                &serde_json::to_value("Wallet not found").unwrap()
-            )
-        }
+        schema
+            .wallet(&query.pub_key)
+            .ok_or_else(|| api::Error::NotFound("Wallet not found".to_owned()))
     }
 
-    fn get_wallets(&self, _: &mut Request) -> IronResult<Response> {
-        let snapshot = self.blockchain.snapshot();
+    /// Endpoint for dumping all wallets from the storage.
+    fn get_wallets(state: &ServiceApiState, _query: ())
+     -> api::Result<Vec<Wallet>> {
+        let snapshot = state.snapshot();
         let schema = CurrencySchema::new(snapshot);
         let idx = schema.wallets();
-        let wallets: Vec<Wallet> = idx.values().collect();
-
-        self.ok_response(&serde_json::to_value(&wallets).unwrap())
+        let wallets = idx.values().collect();
+        Ok(wallets)
     }
 }
 ```
 
 As with the transaction endpoint, the methods have an idiomatic signature
-`fn(&self, &mut Request) -> IronResult<Response>`.
+`fn(&ServiceApiState, MyQueryParam) -> api::Result<MyResponseParam>`.
 
 ### Wire API
 
 As the final step of the API implementation, we need to tie request
-processing logic to
-specific endpoints. We do this in the `CryptocurrencyApi::wire()`
-method:
+processing logic to specific endpoints.
+We do this in the `CryptocurrencyApi::wire()` method:
 
 ```rust
-impl Api for CryptocurrencyApi {
-    fn wire(&self, router: &mut Router) {
-        let self_ = self.clone();
-        let post_create_wallet =
-            move |req: &mut Request| self_.post_transaction(req);
-        let self_ = self.clone();
-        let post_transfer =
-            move |req: &mut Request| self_.post_transaction(req);
-        let self_ = self.clone();
-        let get_wallets = move |req: &mut Request| self_.get_wallets(req);
-        let self_ = self.clone();
-        let get_wallet = move |req: &mut Request| self_.get_wallet(req);
-
-        // Bind handlers to specific routes.
-        router.post("/v1/wallets", post_create_wallet, "post_create_wallet");
-        router.post("/v1/wallets/transfer", post_transfer, "post_transfer");
-        router.get("/v1/wallets", get_wallets, "get_wallets");
-        router.get("/v1/wallet/:pub_key", get_wallet, "get_wallet");
+impl CryptocurrencyApi {
+    fn wire(builder: &mut ServiceApiBuilder) {
+        // Binds handlers to specific routes.
+        builder
+            .public_scope()
+            .endpoint("v1/wallet", Self::get_wallet)
+            .endpoint("v1/wallets", Self::get_wallets)
+            .endpoint_mut("v1/wallets", Self::post_transaction)
```

Comment:
Why two separate methods (`endpoint` and `endpoint_mut`)? That's the only place where they appear in the docs.

Reply:
Single `endpoint` have been splitted to `endpoint` and `endpoint_mut` as result of a long discussion. We need to distinguish write and read requests and thus understand where is a get request and where is a post request in HTTP backend.
I can write more detailed description about the `endpoint` and `endpoint_mut` purposes.

Reply:
Thanks, what the reader of this would benefit most? Maybe, a short explanation of the differences + a link to the API documentation on `docs.rs` (I hope it explains the differences well)?